pycue/tests/cloudinstancegroup_test.py

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python
+
+#  Copyright (c) 2018 Sony Pictures Imageworks Inc.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+
+import mock
+import unittest
+import os
+
+import opencue
+from opencue.cloud.gce_api import GoogleCloudGroup
+from googleapiclient import discovery
+from googleapiclient.http import HttpMock, HttpRequestMock
+
+TEST_CLOUD_GROUP_NAME = "test-group-main"
+PROJECT_NAME = "test-gce-project"
+ZONE_NAME = "test-zone-name"
+GOOGLE_RESOURCES_DIR = os.path.join(os.path.dirname(__file__), "google-test-resources")
+TEST_INSTANCE_TEMPLATE_NAME = "test-instance-template"
+
+
+class GoogleCloudGroupTest(unittest.TestCase):
+
+    def setUp(self):
+        # Use the HttpMock module to build the mock service object
+        self.http = HttpMock(os.path.join(GOOGLE_RESOURCES_DIR, "compute-discovery.json"),
+                             {'status': '200'})
+        api_key = "test_api_key"
+        self.service = discovery.build('compute', 'v1', http=self.http, developerKey=api_key)
```

Comment:
I think that we're including too much third-party Google logic in these tests.

Ideally we don't want to be calling into Google-owned modules at all. We're not testing the cloud provider modules themselves; we only want to make sure our own code is behaving as expected. So we really want to being thinking more along the lines of "if the Google API were to return X as its output, would my code handle that result properly?" So rather than calling `discovery.build` and then mocking out pieces of it, mock the whole class.

This is related to the comment above about global variables. Once the `service` variable is no longer a global, you can use `mock.patch.object` to set up `GoogleCloudGroup.service` as a `Mock` (in lieu of calling `connect()` as the main code does/will do). Then we can mock out `service` as needed without ever calling into those Google modules.

Hopefully that makes sense, can explain more during our next meeting.

Reply:
I have redone the setup for tests for this module. I'll go through these with you during our check-in and then finish up the other tests if this structure looks good